vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -237,15 +237,30 @@ import { isFocusable } from '@vaadin/grid/src/vaadin-grid-active-item-mixin.js';
         });
         grid._createPropertyObserver('activeItem', '__activeItemChangedDetails', true);
 
-        grid.$connector._getPageIfSameLevel = tryCatchWrapper(function (parentKey, index, defaultPage) {
-          let cacheAndIndex = grid._cache.getCacheAndIndex(index);
-          let parentItem = cacheAndIndex.cache.parentItem;
-          let parentKeyOfIndex = parentItem ? grid.getItemId(parentItem) : root;
-          if (parentKey !== parentKeyOfIndex) {
-            return defaultPage;
-          } else {
-            return grid._getPageForIndex(cacheAndIndex.scaledIndex);
+        grid.$connector._getSameLevelPage = tryCatchWrapper(function (
+          parentKey,
+          potentialSameLevelCache,
+          potentialSameLevelItemKey
+        ) {
+          if (!potentialSameLevelCache) {
+            // There is no parent cache to match level
+            return null;
+          }
+          const parentKeyOfPotentialSameLevelCache = potentialSameLevelCache.parentItem
+            ? grid.getItemId(potentialSameLevelCache.parentItem)
+            : root;
+          if (parentKeyOfPotentialSameLevelCache === parentKey) {
+            // Level match found
+            return grid._getPageForIndex(grid._cache.getCacheAndIndexByKey(potentialSameLevelItemKey).scaledIndex);
```

Comment:
It would be good to avoid calling `getCacheAndIndexByKey` in this case as well. Once we determined the cache to which the desired item belongs, we can simply iterate through `cache.items` to locate its scaled index.